crates/ibc/src/host_helpers.rs

```diff
@@ -0,0 +1,112 @@
+                Error::invalid_client_state("client must be a tendermint client".to_string())
+            })?;
+
+        if counterparty_client_state.is_frozen() {
+            return Err(Error::invalid_client_state("client is frozen".to_string()));
+        }
+
+        let self_chain_id = self.chain_id();
+        if self_chain_id != &counterparty_client_state.chain_id {
+            return Err(Error::invalid_client_state(format!(
+                "invalid chain-id. expected: {}, got: {}",
+                self_chain_id, counterparty_client_state.chain_id
+            )));
+        }
+
+        let self_revision_number = self_chain_id.version();
+        if self_revision_number != counterparty_client_state.latest_height().revision_number() {
+            return Err(Error::invalid_client_state(format!(
+                "client is not in the same revision as the chain. expected: {}, got: {}",
+                self_revision_number,
+                counterparty_client_state.latest_height().revision_number()
+            )));
+        }
+
+        if counterparty_client_state.latest_height() >= ctx.host_current_height() {
+            return Err(Error::invalid_client_state(format!(
+                "client has latest height {} greater than or equal to chain height {}",
+                counterparty_client_state.latest_height(),
+                ctx.host_current_height()
+            )));
+        }
+
+        if self.proof_specs() != &counterparty_client_state.proof_specs {
+            return Err(Error::invalid_client_state(format!(
+                "client has invalid proof specs. expected: {:?}, got: {:?}",
+                self.proof_specs(),
+                counterparty_client_state.proof_specs
+            )));
+        }
+
+        let _ = {
+            let trust_level = counterparty_client_state.trust_level;
+
+            TendermintTrustThresholdFraction::new(
+                trust_level.numerator(),
+                trust_level.denominator(),
+            )
+            .map_err(|_| Error::invalid_client_state("invalid trust level".to_string()))?
+        };
+
+        if self.unbonding_period() != counterparty_client_state.unbonding_period {
+            return Err(Error::invalid_client_state(format!(
+                "invalid unbonding period. expected: {:?}, got: {:?}",
+                self.unbonding_period(),
+                counterparty_client_state.unbonding_period,
+            )));
+        }
+
+        if counterparty_client_state.unbonding_period < counterparty_client_state.trusting_period {
+            return Err(Error::invalid_client_state(format!(
+                "unbonding period must be greater than trusting period. unbonding period ({:?}) < trusting period ({:?})",
+                counterparty_client_state.unbonding_period,
+                counterparty_client_state.trusting_period
+            )));
+        }
+
+        if self.upgrade_path() != counterparty_client_state.upgrade_path {
```

Comment:
Just noticed that ibc-go allows an empty `counterparty_client_state.upgrade_path` regardless of whether or not the host upgrade path is empty -> https://github.com/cosmos/ibc-go/blob/v5.0.1/modules/core/02-client/keeper/keeper.go#L322-L329

Reply:
Great catch! [af79e61](https://github.com/cosmos/ibc-rs/pull/207/commits/af79e611b108ee4644b184a82007dc5dd24e0e34)